### PR TITLE
hackathon-2024-03-watch-out

### PR DESCRIPTION
**Pull request by diegommm**
**What is this feature?**

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/store/entity/sqlstash/sql_storage_server.go [hunk 1]
```diff
@@ -1486,8 +1582,345 @@ func watchMatches(r *entity.EntityWatchRequest, result *entity.Entity) bool {
...
+			}
+
+			// update r.Since value so we don't send earlier results again
+			r.Since = result.ResourceVersion
+		}
+	}
+
+	return nil
+}
+
```
Comment (DanCech) [inline pkg/services/store/entity/sqlstash/sql_storage_server.go:1921]: there seems to be a lot of duplication between this code and the existing poll-based setup, is there a reason not to just update line ~147 to choose between the standard poller and the pg watch implementation, but keep the same broadcast code?

File: pkg/services/store/entity/tests/server_integration_test.go [hunk 2]
```diff
@@ -109,6 +112,84 @@ func requireVersionMatch(t *testing.T, obj *entity.Entity, m objectVersionMatche
...
+				Name:      name,
+				Body:      body,
+				Message:   "first entity!",
+			},
+		}
+		_, err = testCtx.client.Create(testCtx.ctx, createReq)
+		require.NoError(t, err)
+
+		// watch client receives create
+		res, err := watchClient.Recv()
```
Comment (owensmallwood) [inline pkg/services/store/entity/tests/server_integration_test.go:166]: Sometimes this hangs forever... sometimes it doesn't 🤯

File: pkg/services/store/entity/sqlstash/sql_storage_server.go (hunk 1 above)
Comment (diegommm) [reply pkg/services/store/entity/sqlstash/sql_storage_server.go]: Actually, I think I can work around it to keep track of the postgres channels listened to separately and unlisten and terminate separately, and that would make the code more compatible with Broadcaster. I'll try to do that, but I'd like to wrap a bit first for the hackathon video as well. But I also think we may need the unbounded buffer channel implementation for the reasons above, wdyt?

**Comment by diegommm** [comment]
Closing as related work is now under development in https://github.com/grafana/grafana/pull/84629 and other PRs.

**Pull request closed without merging**